File: Client/Android/app/src/main/java/com/photomanager/photomanager/main/home/api/Geolocation.kt
package com.photomanager.photomanager.main.home.api

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class Geolocation(
    @SerialName("latitude") val latitude: Double,
    @SerialName("longitude") val longitude: Double
)

@Serializable
data class Place(
    @SerialName("name") val name: String,
    @SerialName("aliases") val aliases: List<String>,
    @SerialName("city") val city: String,
    @SerialName("country") val country: String
)

File: Server/models/geolocation.go
package models

import (
	"go.mongodb.org/mongo-driver/bson"
)

type Geolocation struct {
	Latitude  float64 `json:"latitude"`
	Longitude float64 `json:"longitude"`
}

type Place struct {
	Name    string   `json:"name" bson:"name"`
	Aliases []string `json:"aliases" bson:"aliases"`
	City    string   `json:"city" bson:"city"`
	Country string   `json:"country" bson:"country"`
}

type geoJson struct {
	Type        string    `bson:"type"`
	Coordinates []float64 `bson:"coordinates"`
}

func (g *Geolocation) MarshalBSON() ([]byte, error) {
	marshalled, err := bson.Marshal(geoJson{"Point", []float64{g.Longitude, g.Latitude}})
	return marshalled, err
}

func (g *Geolocation) UnmarshalBSON(data []byte) error {
	unmarshalled := geoJson{}
	err := bson.Unmarshal(data, &unmarshalled)
	if err != nil {
		return err
	}
	g.Latitude = unmarshalled.Coordinates[1]
	g.Longitude = unmarshalled.Coordinates[0]
	return nil
}
